cupy/util.pyx

```diff
@@ -25,6 +26,35 @@ except AttributeError:  # python <3.3
 cdef list _memos = []
 
 
+def _normalize_axis_index(axis, ndim):
```

Comment:
Implementation of this function is not correct. Note that `a <= b < c` is equivalent to `a <= b and b < c`.
How about just copy from this code?
https://github.com/cupy/cupy/blob/678466d06e237d28cc18fd483e15d71771d3f8a3/cupy/core/_routines_sorting.pyx#L83-L86

Reply:
> Note that `a <= b < c` is equivalent to `a <= b and b < c`
That's the expected behaviour. That's what `_normalize_axis_index` checks for.

Reply:
For example, if `axis = -100` and `ndim = 3`, the expression `-axis <= ndim < axis` evaluates to `False`, while we expect an error.

Reply:
Fixed it

Reply:
Thank you for the fix, but I think the above four lines are sufficient for this function.
How do you think?

Reply:
I'm not sure if it'll work because the above lines adds `ndim` to `_axis` if `_axis < 0`, but `_normalize_axis_error` checks if `ndim` is in `[-axis,axis)`

Reply:
Do you mean `axis` is in `[-ndim, ndim)`?
Let's divide the domain into four sections, split at `-ndim`, 0 and `ndim`:

* If `axis < -ndim`: `axis + ndim < 0` => `False`
* If `-ndim <= axis < 0`: `0 <= axis + ndim < ndim`  => `True`
* If `0 <= axis < ndim`: => directly `True`
* If `ndim <= axis`: directly `False`

So, I think this logic is correct.

Reply:
@niboshi Yes this logic is right. I have changed it.

Reply:
Thank you for the fix.
Sorry, maybe I was ambiguous, but I think the whole implementation of this function is written with just these lines:

```py
if axis < 0: 
    axis += ndim 
if not (0 <= axis < ndim): 
    raise _errors._AxisError(...) 
return axis
```

Reply:
@niboshi Yes I have changed the definition to be simple. Have a look

Reply:
@niboshi Yes it makes more sense now. I have changed the definition to be simple. Have a look ;)